packages/protobuf/src/proto.ts

```diff
@@ -0,0 +1,206 @@
+    In: StreamingMode.In,
+    Out: StreamingMode.Out,
+    None: StreamingMode.None,
+  }[mode.name as string];
+
+  ctx.program.stateMap(state.stream).set(target, emitStreamingMode);
+}
+
+function getTuple(program: Program, t: Type): [number, number] | null {
+  if (t.kind !== "Tuple" || t.values.some((v) => v.kind !== "Number") || t.values.length !== 2) {
+    reportDiagnostic(program, {
+      code: "illegal-reservation",
+      target: t,
+    });
+
+    return null;
+  }
+
+  return Object.assign(
+    (t as Tuple).values.map((v) => (v as NumericLiteral).value) as [number, number],
+    { type: t }
+  );
+}
+
+export type Reservation = string | number | ([number, number] & { type: Type });
+
+export function $reserve(
+  ctx: DecoratorContext,
+  target: Model,
+  ...reservations: readonly (Type | number | string)[]
+) {
+  const finalReservations = reservations
+    .map((reservation) =>
+      typeof reservation === "object" ? getTuple(ctx.program, reservation) : reservation
+    )
+    .filter((v) => v != null);
+
+  ctx.program.stateMap(state.reserve).set(target, finalReservations);
+}
+
+export function $message(ctx: DecoratorContext, target: Model) {
+  ctx.program.stateSet(state.message).add(target);
+}
+
+/**
+ * Decorate a model property with a field index. Field indices are required for all fields of emitted messages.
+ *
+ * @param param0
+ * @param target
+ * @param fieldIndex
+ * @returns
+ */
+export function $field(ctx: DecoratorContext, target: ModelProperty, fieldIndex: number) {
+  if (!Number.isInteger(fieldIndex) || fieldIndex <= 0) {
+    reportDiagnostic(ctx.program, {
+      code: "field-index",
+      messageId: "invalid",
+      format: {
+        index: String(fieldIndex),
+      },
+      target,
+    });
+    return;
+  } else if (fieldIndex > MAX_FIELD_INDEX) {
+    reportDiagnostic(ctx.program, {
+      code: "field-index",
+      messageId: "out-of-bounds",
+      format: {
+        index: String(fieldIndex),
+        max: String(MAX_FIELD_INDEX + 1),
+      },
+      target,
+    });
+    return;
+  } else if (
+    fieldIndex >= IMPLEMENTATION_RESERVED_RANGE[0] &&
+    fieldIndex <= IMPLEMENTATION_RESERVED_RANGE[1]
+  ) {
+    reportDiagnostic(ctx.program, {
+      code: "field-index",
+      messageId: "reserved",
+      format: {
+        index: String(fieldIndex),
+      },
+      target,
+    });
+  }
+
+  ctx.program.stateMap(state.fieldIndex).set(target, fieldIndex);
+}
+
+/**
+ * Emitter main function.
+ *
+ * @param program - the program to emit
+ */
+export async function $onEmit(ctx: EmitContext<EmitOptionsFor<TypeSpecProtobufLibrary>>) {
+  const emitter = createProtobufEmitter(ctx.program);
+
+  await emitter(resolvePath(ctx.emitterOutputDir), ctx.options);
+}
+
+/**
+ * Validation function
+ */
+export async function $onValidate(program: Program) {
+  /* c8 ignore next 6 */
+  if (program.compilerOptions.noEmit) {
+    const options = program.emitters.find((e) => e.emitFunction === $onEmit)
```

Comment:
Here should ideally not have to check the emitter options for onValidate. THe code should ideally be able to be validated the same regardless of what is the emitter configuration and then if there is some failure specific to that its probably not worth returning in onValidate.(e.g. FIlename conflict)

Reply:
I've created an issue to track this: https://github.com/microsoft/typespec/issues/1859. I've also added a link to this issue in the source.

The core problem is that without checking that the `noEmit` option is set, I generate double diagnostics, since all diagnostics come from the emitter.